Fix the issue of TextSelection could misbehave inside a Nested Overlay

shihchanghsiungsonos opened the pull request:
## Description

- Fixes the alignment issue and the touch issues.
- Uses the screen size for laying out the toolbar. so that it doesn't use the MediaQuery size of the screen, but instead the bounding size of the overlay the toolbar is added to. This allows the toolbar to be correctly aligned with the right side of our card.

## Related Issues

https://github.com/flutter/flutter/issues/43787

## Tests

Verify the issue is fixed and no build break.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

dannyvalentesonos commented:
@justinmc here's the PR for fixing the aligning and tap issues with a nested Overlay. The overflow issue is logged separately.
Thanks!

dnfield commented on packages/flutter/lib/src/material/text_selection.dart:
```diff
@@ -162,11 +162,13 @@ class _MaterialTextSelectionControls extends TextSelectionControls {
     final Offset preciseMidpoint = Offset(position.dx, y);
 
+    // get the Overlay's render box
+    RenderBox overlayBox = Overlay.of(context).context?.findRenderObject();
```
Why `context?`?  

If that's actually intended, then we should handle the case where this is null below - I'm not super familiar with the overlay API so I'm not sure if that's something we need to be concerned with though.

dnfield replied:
Also, this should be `final`.

dannyvalentesonos replied:
Thanks @dnfield for the review. Yes, I was thinking about this same thing over the weekend actually. I don't think the context would be null if we can actually find it with the `.of` method, but I wasn't sure if we could always assume that. I was going to make the next line below be `overlayBox?.size ?? MediaQuery.of(context).size` but I'd rather wait to get more feedback on whether this is actually needed or not.

dnfield replied:
I'm pretty sure you would always have the context in this method, and if not it'd be a bug.  The context is just basically the `Element` for the widget, so if you can't find your parent element you're in trouble 😆 

We could add an `assert` to that effect - e.g. 
```dart
final OverlayState overlayState = Overlay.of(context);
assert(overlayState != null);
assert(overlayState.context != null);
```

dnfield commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -528,8 +531,8 @@ class TextSelectionOverlay {
     final Rect editingRegion = Rect.fromPoints(
-      renderObject.localToGlobal(Offset.zero),
-      renderObject.localToGlobal(renderObject.size.bottomRight(Offset.zero)),
+      renderObject.localToGlobal(Offset.zero, ancestor: _overlay.context?.findRenderObject()),
```
Ditto on the `?`s on these two lines.  We need to handle the case where these are null if they actually can be null.

Also, this method gets called before you set `_overlay` above. What if `_overlay` is null?

dannyvalentesonos replied:
We assert that `_overlay` is not null, and it's set in the constructor. Can you please explain how this method gets called before the constructor?
The reason I added `_overlay` is so that we don't do a lookup every time `_buildToolbar` is called, although maybe that's not much of a concern seeing this should not be called every frame.

Finally, on your comment about it being null... if it's null, then the `localToGlobal`'s `ancestor` parameter would be null, which means the global bounds will be used, which I think is OK in that case.

Maybe we just change it so that we call Overlay.of in this method and not save the `_overlay` member?

shihchanghsiungsonos replied:
@dnfield Thanks for the review. The PR has been updated to address your comments.

dnfield commented:
This will need some tests.  As I mentioned, I'm not super familiar with the overlay API but someone else who is can comment more on that.

justinmc reviewed (commented):
I tried this locally and found one case that we should fix:

<img width="168" alt="Screen Shot 2019-11-04 at 10 57 05 AM" src="https://user-images.githubusercontent.com/389558/68148986-eab7d380-fef1-11e9-8da7-52fd1e67a841.png">

If you make the Overlay more narrow than the selection menu, then you will get overflow.  I think ideally the selection menu should be interactive outside of the Overlay, though that might not be possible...    If not, maybe we'll need horizontal scrolling in the selection menu to handle this edge case.  This solution did seem to work fine in normal cases for the record.

Also, we will definitely need at least one test to show that a selection menu inside of an Overlay is useable.

dannyvalentesonos commented on packages/flutter/lib/src/material/text_selection.dart:
```diff
@@ -162,11 +162,13 @@ class _MaterialTextSelectionControls extends TextSelectionControls {
     final Offset preciseMidpoint = Offset(position.dx, y);
 
+    // get the Overlay's render box
+    final RenderBox overlayBox = Overlay.of(context).context?.findRenderObject();
```
Thanks @justinmc . Yes the overflow is an issue, even without using a Nested Overlay or even making any overlay smaller. Just using the LargeDisplay setting on the OS will cause this. I've logged #44005. I think that needs to be handled exactly the same way the OS handles it, with the overflow menu.

justinmc replied:
Good point, that sounds good to me.  Thanks for opening the issue.

dannyvalentesonos replied:
Also, just to add my opinion. After testing this for days, I went back and forth between allowing it to display outside the containing overlay, which would require a different fix to get the toolbar to be interact-able, and having it laid out inside its containing overlay. I have to say, that it actually feels much more natural and "expected" when it's aligned nicely with its containing overlay.

justinmc commented:
@goderbauer We could use some of your Overlay expertise if you have time to give a review!

goderbauer commented on packages/flutter/lib/src/material/text_selection.dart:
```diff
@@ -162,11 +162,17 @@ class _MaterialTextSelectionControls extends TextSelectionControls {
         : startTextSelectionPoint.point.dy + _kToolbarHeight + _kToolbarContentDistanceBelow;
     final Offset preciseMidpoint = Offset(position.dx, y);
 
+    // get the Overlay's render box
```
That the toolbar is placed in an overlay seems like an implementation detail of the `TextSelectionOverlay` class. Do we really want to have implementations of TextSelectionControls depend on that? Maybe `buildToolbar` should take some kind of a constraints parameter instead specifying how big the tool bar can be?

dannyvalentesonos replied:
I can look at doing this.
And to answer your larger picture question of whether there's a reason to have the toolbar in the child overlay... I think it makes more sense that way, as the toolbar will not start floating around far outside the actual child overlay's bounds, which could look kind of weird. After trying both ways for a few days, I can say that having it in the child overlay as it already is, and with the fix being presented here, it feels much more natural.

dannyvalentesonos replied:
@goderbauer Actually, I'm not sure if I should really be modifying the `TextSelectionControls`'s interface, since I imagine there can be different implementations of it for Web/desktop, etc that I'm not aware of, nor do I have any way of testing it. I'd be happy to have someone else take over from here, if that works.

dannyvalentesonos replied:
@goderbauer Any suggestions on this? Are the material and cupertino text_selection.dart files the only ones that implement that API? If so, I can make that change too.

goderbauer reviewed (commented):
Looking at this I am also wondering if placing the SelectionControlls into the nearest overlay was the right choice. Maybe these should actually go in the root overlay, which usually covers the entire screen? Is there any good reason we need to contain them in a child Overlay?

shihchanghsiungsonos commented:
> @googlebot I fixed it.

shihchanghsiungsonos commented:
@googlebot I fixed it.

shihchanghsiungsonos commented:
@googlebot I fixed it.

shihchanghsiungsonos commented:
> We found a Contributor License Agreement for you (the sender of this pull request), but were unable to find agreements for all the commit author(s) or Co-authors. If you authored these, maybe you used a different email address in the git commits than was used to sign the CLA ([login here](https://cla.developers.google.com/) to double check)? If these were authored by someone else, then they will need to sign a CLA as well, and confirm that they're okay with these being contributed to Google.
> In order to pass this check, please resolve this problem and then comment `@googlebot I fixed it.`. If the bot doesn't comment, it means it doesn't think anything has changed.
> 
> ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Fflutter%2Fflutter%2Fpull%2F44018) for more info**.

@googlebot I fixed it.

justinmc reviewed (commented):
I like this solution, I think it's just missing a test now if others agree.  I think you can add your test to material/text_field_test.dart.  It will be similar to [this test](https://github.com/flutter/flutter/blob/bc6f270c584d1fdba81330090ef6e822b9082919/packages/flutter/test/material/text_field_test.dart#L1578).  I'd put a text field into a narrow Container and expect that all 4 text selection menu buttons are still there.  Just make sure your test would fail without these changes.

justinmc commented:
@goderbauer @dnfield for a re-review.

dannyvalentesonos commented on packages/flutter/lib/src/widgets/overlay.dart:
```diff
@@ -233,8 +233,18 @@ class Overlay extends StatefulWidget {
+  /// above all subsequent instances of [Overlay].
+  static OverlayState of(
+    BuildContext context, {
+    bool rootOverlay = false,
```
I followed the pattern in navigator.dart which uses `rootNavigator` as the variable name.

justinmc replied:
Ah ok, that sounds like a good reason to keep it as rootOverlay to me :+1:

dnfield replied:
I would actually make this a separate static method on `Overlay`, e.g.

```dart
static Overlay rootOf(BuildContext context) { ... }
```

The convention is that `of` finds the closest ancestor of this type.

dannyvalentesonos replied:
I am not against that, and will change it if most feel that's the right thing. I'm just pasting here what the navigator's version looks like
```
static NavigatorState of(
    BuildContext context, {
    bool rootNavigator = false,
    bool nullOk = false,
  }) {
```

dnfield replied:
Ah in that case this seems fine then.

dnfield replied:
This method should have a test making sure it works properly :)

dnfield commented:
This seems to make sense to me - my comment is main a nit, I'd be interested in what @goderbauer thinks of the naming.  Or @HansMuller who has done a lot with these kinds of things.

goderbauer commented:
This looks fine. It is missing a test, though and the CLA needs to be fixed.

dannyvalentesonos commented:
> I like this solution, I think it's just missing a test now if others agree. I think you can add your test to material/text_field_test.dart. It will be similar to [this test](https://github.com/flutter/flutter/blob/bc6f270c584d1fdba81330090ef6e822b9082919/packages/flutter/test/material/text_field_test.dart#L1578). I'd put a text field into a narrow Container and expect that all 4 text selection menu buttons are still there. Just make sure your test would fail without these changes.

@justinmc I've added the following test, but it's failing to find the 'SELECT ALL' text entirely when the Container has a small size set.
```
testWidgets(
    'Check the toolbar appears correctly inside a nested overlay',
    (WidgetTester tester) async {
      // This is a regression test for
      // https://github.com/flutter/flutter/issues/43787
      final TextEditingController controller = TextEditingController();

      await tester.pumpWidget(MaterialApp(
        home: Scaffold(
          body: Center(
            child: Container(
              width: 300,
              height: double.infinity,
              child: Overlay(
                initialEntries: [
                  OverlayEntry(
                      builder: (context) => Material(
                        child: Padding(
                          padding: const EdgeInsets.all(30.0),
                          child: TextField(
                            controller: controller,
                          ),
                        ),
                      ),
                  ),
                ],
              ),
            ),
          ),
        ),
      ));

      const String testValue = 'abc def ghi';
      await tester.enterText(find.byType(TextField), testValue);
      await skipPastScrollingAnimation(tester);

      await _showSelectionMenuAt(tester, controller, testValue.indexOf('e'));

      // Verify the selection toolbar position is below the text.
      Offset toolbarTopLeft = tester.getTopLeft(find.text('SELECT ALL'));
      Offset textFieldTopLeft = tester.getTopLeft(find.byType(TextField));
      expect(textFieldTopLeft.dy, lessThan(toolbarTopLeft.dy));
    },
  );
```

If I change the width to 500, it finds it.

I've added printing in a few places in the sdk, and the following outputs can be seen during the test:

```
00:04 +1: Check the toolbar appears correctly inside a nested overlay                                                                                            
media: MediaQueryData(size: Size(800.0, 600.0), devicePixelRatio: 3.0, textScaleFactor: 1.0, platformBrightness: Brightness.light, padding: EdgeInsets.zero, viewPadding: EdgeInsets.zero, viewInsets: EdgeInsets.zero, physicalDepth: 1.7976931348623157e+308, alwaysUse24HourFormat: false, accessibleNavigation: false, highContrast: false,disableAnimations: false, invertColors: false, boldText: false)
rootOverlay: true
rootOverlay: true
rootOverlay: true
rootOverlay: true
rootOverlay: true
rootOverlay: true
rootOverlay: true
editingRegion: Rect.fromLTRB(280.0, 46.0, 520.0, 62.0)
opening toolbar at Offset(143.0, 78.0) with size: Size(450.0, 44.0)
```

All these numbers seem good, except for the toolbar width of (450), which is wrong. This log was added in the `_TextSelectionToolbarLayout`'s `getPositionForChild` method's given `childSize`.

Other tests show the output as :
```
opening at Offset(13.0, 114.0) with size: Size(274.0, 44.0)
```

So because of this, something isn't working right with the test, yet running an actual app with the same tree works well on an Android device.

I think I might need some help with figuring out if anything else is missing in my fix.

Thanks

dannyvalentesonos commented:
@justinmc 
Interesting. If I change this test from
```
testWidgets(
    'Check the toolbar appears below the TextField when there is not enough space above the TextField to show it',
    (WidgetTester tester) async {
      // This is a regression test for
      // https://github.com/flutter/flutter/issues/29808
      final TextEditingController controller = TextEditingController();

      await tester.pumpWidget(MaterialApp(
        home: Scaffold(
          body: Padding(
            padding: const EdgeInsets.all(30.0),
            child: TextField(
              controller: controller,
              ),
            ),
          ),
        ),
      );

      const String testValue = 'abc def ghi';
      await tester.enterText(find.byType(TextField), testValue);
      await skipPastScrollingAnimation(tester);

      await _showSelectionMenuAt(tester, controller, testValue.indexOf('e'));

      // Verify the selection toolbar position is below the text.
      Offset toolbarTopLeft = tester.getTopLeft(find.text('SELECT ALL'));
      Offset textFieldTopLeft = tester.getTopLeft(find.byType(TextField));
      expect(textFieldTopLeft.dy, lessThan(toolbarTopLeft.dy));

      await tester.pumpWidget(MaterialApp(
        home: Scaffold(
          body: Padding(
            padding: const EdgeInsets.all(150.0),
            child: TextField(
              controller: controller,
            ),
          ),
        ),
      ));

      await tester.enterText(find.byType(TextField), testValue);
      await skipPastScrollingAnimation(tester);

      await _showSelectionMenuAt(tester, controller, testValue.indexOf('e'));

      // Verify the selection toolbar position
      toolbarTopLeft = tester.getTopLeft(find.text('SELECT ALL'));
      textFieldTopLeft = tester.getTopLeft(find.byType(TextField));
      expect(toolbarTopLeft.dy, lessThan(textFieldTopLeft.dy));
    },
  );
```
to
```
testWidgets(
    'Check the toolbar appears below the TextField when there is not enough space above the TextField to show it',
    (WidgetTester tester) async {
      // This is a regression test for
      // https://github.com/flutter/flutter/issues/29808
      final TextEditingController controller = TextEditingController();

      await tester.pumpWidget(MaterialApp(
        home: Scaffold(
          body: Center(
            child: Container(
              width: 300,
              child: Padding(
                padding: const EdgeInsets.all(30.0),
                child: TextField(
                  controller: controller,
                ),
              ),
            ),
          ),
        ),
      ));

      const String testValue = 'abc def ghi';
      await tester.enterText(find.byType(TextField), testValue);
      await skipPastScrollingAnimation(tester);

      await _showSelectionMenuAt(tester, controller, testValue.indexOf('e'));

      // Verify the selection toolbar position is below the text.
      Offset toolbarTopLeft = tester.getTopLeft(find.text('SELECT ALL'));
      Offset textFieldTopLeft = tester.getTopLeft(find.byType(TextField));
      expect(textFieldTopLeft.dy, lessThan(toolbarTopLeft.dy));
      
    },
  );
```

I get the same error, with the same output. So it actually has nothing to do with the nested overlay, and this case should really work.

Is there something I'm missing in terms of this test code?

Thanks

justinmc commented:
I tried running the PR to see what could be happening in these tests, but I noticed that it still seems broken to me in the Android emulator.  I'm just using the code from the original issue.

<img width="380" alt="Screen Shot 2020-01-15 at 11 46 52 AM" src="https://user-images.githubusercontent.com/389558/72466130-26daa080-378d-11ea-88b3-b342928c761a.png">

dannyvalentesonos commented:
Hmm, that's strange, because it doesn't look like the new version of the fix is being applied. Is it possible you have the first version I posted? That's what it looks like.

shihchanghsiungsonos commented:
> This looks fine. It is missing a test, though and the CLA needs to be fixed.

I have tried to fix the CLA issue many times, but no luck. I also went through all my Google accounts. They all show `It looks like you've already signed this CLA`. Can anyone know how to find out what is missing? I also have other PR and it shows CLA been signed. Thanks!

dnfield commented:
@shihchanghsiungsonos  - when I look up your GitHub user name it doesn't come up in our CLA database :(  Are you sure you've signed it with the information associated with this GitHub account?

shihchanghsiungsonos commented:
> @shihchanghsiungsonos - when I look up your GitHub user name it doesn't come up in our CLA database :( Are you sure you've signed it with the information associated with this GitHub account?

@dnfield thanks for looking into this. My Google account is `shihchang.hsiung@sonos.com` and I use that to sign in GitHub account.

dnfield commented:
That email comes up as green.  And I can see from the patch that your email is the right one.  I'm going to manually mark this as good.

dnfield commented:
Ah! I see now - @shihchanghsiungsonos - can you go to https://cla.developers.google.com/clas and update your `GitHub Username` to match your actual username?  It looks like you have your email as your username instead of your username :)

shihchanghsiungsonos commented:
> shihchanghsiungsonos

Done. Thanks a lot for the help, @dnfield ! :)

justinmc commented:
@dannyvalentesonos You're right that I was on an older version, sorry about that!

Actually I think you can do something simpler for the test and just test that there is no error when rendering inside of an overlay.  Something like this seems to fail on master but pass on your branch:

```dart
  testWidgets(
    'The selection menu displays in an Overlay without error',
    (WidgetTester tester) async {
      // This is a regression test for
      // https://github.com/flutter/flutter/issues/43787
      final TextEditingController controller = TextEditingController(
        text: 'This is a test that shows some odd behavior with Text Selection!',
      );

      await tester.pumpWidget(MaterialApp(
        home: Scaffold(
          body: Container(
            color: Colors.grey,
            child: Center(
              child: Container(
                color: Colors.red,
                width: 300,
                height: 600,
                child: Overlay(
                  initialEntries: [
                    OverlayEntry(
                      builder: (context) => Center(
                        child: TextField(
                          controller: controller,
                        ),
                      ),
                    )
                  ],
                ),
              ),
            ),
          ),
        ),
      ));

      await _showSelectionMenuAt(tester, controller, controller.text.indexOf('test'));
      await tester.pumpAndSettle();
      expect(tester.takeException(), isNull); 
    },
  );
```

dannyvalentesonos commented:
Awesome, thanks @justinmc . We'll get the PR updated.

dannyvalentesonos commented:
@justinmc An update to the PR has been made. Can you take a look please?
I'm not sure why flutter-build is broken... Is that something in our changes?

justinmc reviewed (approved):
LGTM

The failures look strange.  @dannyvalentesonos Could you try merging master?

dannyvalentesonos commented:
Thanks @justinmc . We updated. flutter-build is passing, but not sure about all the other tests.

justinmc commented:
@dannyvalentesonos Weird failures are still happening.  I talked to someone on the Flutter discord and they said to run `git reset --soft && git commit -am 'message' && git push -f`.  Something in the tool messed up while trying to read the versions.

Hopefully that fixes things!

shihchanghsiungsonos commented:
> @dannyvalentesonos Weird failures are still happening. I talked to someone on the Flutter discord and they said to run `git reset --soft && git commit -am 'message' && git push -f`. Something in the tool messed up while trying to read the versions.
> 
> Hopefully that fixes things!

After running `git push -f`, looks like the PR has been merged. Not sure if it is right?

dnfield commented:
You should be able to re-open...

It has not been merged.

dnfield commented:
Worst case scenario, push it to a new branch and open a new PR.

dnfield commented:
Ah, it looks like whatever you ended up doing dropped your changes....  see if you can just recover them and push it to a new branch.

shihchanghsiungsonos commented:
> Ah, it looks like whatever you ended up doing dropped your changes.... see if you can just recover them and push it to a new branch.

Thanks @dnfield

shihchanghsiungsonos commented:
> > Ah, it looks like whatever you ended up doing dropped your changes.... see if you can just recover them and push it to a new branch.
> 
> Thanks @dnfield

I have created a new PR: https://github.com/flutter/flutter/pull/49257 Thanks!
